Codefriar/crud fls updates

codefriar opened the pull request:
@pozil - This PR is at the request of @ccoenraets, who's cc'd for visibility. This PR does two main things:

1. it brings Prettier, and npm scripts to the *old* dreamhouse-sfdx project. 
   * The Pretterrc and prettier ignore files have been pulled from apex-recipes. 
   * .gitIgnore has been copied from apex-recipes
   * package.json has been copied from, and modified from dreamhouse-lwc
2. It brings requested security updates discovered by the CLI scanner team. Specifically: 
   * All queries now utilize 'WITH SECURITY_ENFORCED'
   * SampleDataController has been refactored with a guard pattern to fail-fast if the executing user does not have proper CRUD permissions to the necessary objects. It's also been refactored to reduce complexity and adhere to best-practices.

Commit:
adding git ignore file

Commit:
this (massive) commit imposes contemporary
prettier config and prettier ignore setup.

Commit:
Refactors:
PropertyController: now includes WITH SECURITY_ENFORCED on all queries
SampleDataController: refactored
 * with guard pattern to fail-fast on CRUD permission issues
 * to reduce cyclomatic complexity and adhere to best practices.

pozil commented on force-app/main/default/triggers/RejectDuplicatePropertyFavorite.trigger:
```diff
@@ -1,11 +1,16 @@
+        WHERE
+            Property__c = :favorite.Property__c
+            AND User__c = :favorite.User__c
+    ];
```
Enforce FLS?

pozil commented on force-app/main/default/classes/HandlerMyOpenCases.cls:
```diff
@@ -1,25 +1,56 @@
+                Contact.Id,
+                Contact.Name
+            FROM Case
+            WHERE OwnerId = :UserInfo.getUserId() AND Status != 'Closed'
```
Enforce FLS?

pozil commented on force-app/main/default/classes/HandlerNext.cls:
```diff
@@ -1,43 +1,95 @@
+            FROM Opportunity
+            WHERE isClosed = FALSE
+            ORDER BY amount DESC
+            LIMIT 1
```
Enforce FLS?

pozil commented on force-app/main/default/classes/HandlerNext.cls:
```diff
@@ -1,43 +1,95 @@
+            WHERE
+                OwnerId = :UserInfo.getUserId()
+                AND Priority = 'High'
+                AND Status != 'Closed'
```
Enforce FLS?

pozil commented on force-app/main/default/classes/HandlerReport.cls:
```diff
@@ -1,25 +1,45 @@
+            FROM Report
+            WHERE Name LIKE :key
+            ORDER BY Name
+            LIMIT 5
```
Enforce FLS?

pozil commented on force-app/main/default/classes/HandlerSOQL.cls:
```diff
@@ -1,32 +1,49 @@
+        String fileName,
+        String fileContent
+    ) {
         SObject[] objects = Database.query(utterance);
```
We need to figure out what goes in here and enforce FLS.

pozil commented on force-app/main/default/classes/HandlerTopOpportunities.cls:
```diff
@@ -1,22 +1,46 @@
+            FROM Opportunity
+            WHERE isClosed = FALSE
+            ORDER BY amount DESC
+            LIMIT :qty
```
Enforce FLS?

pozil commented on force-app/main/default/classes/PostPriceChangeToSlack.cls:
```diff
@@ -1,54 +1,68 @@
+        Property__c property = [
+            SELECT Address__c, City__c, State__c, Price__c
+            FROM Property__c
+            WHERE Id = :propId
```
Enforce FLS?

pozil commented on force-app/main/default/classes/PushPriceChangeNotification.cls:
```diff
@@ -1,47 +1,58 @@
+        Property__c property = [
+            SELECT Name, Price__c
+            FROM Property__c
+            WHERE Id = :propId
```
Enforce FLS?

pozil commented on force-app/main/default/classes/PushPriceChangeNotification.cls:
```diff
@@ -1,47 +1,58 @@
+        List<Property_Favorite__c> favorites = [
+            SELECT user__c
+            FROM Property_Favorite__c
+            WHERE property__c = :propId
```
Enforce FLS?

pozil commented on force-app/main/default/classes/SlackOpportunityPublisher.cls:
```diff
@@ -1,43 +1,50 @@
+        Opportunity opportunity = [
+            SELECT Name, StageName
+            FROM Opportunity
+            WHERE Id = :oppId
```
Enforce FLS?

pozil reviewed (changes requested):
@codefriar I found a few queries that might require FLS enforcement, can you double check?
Sorry for the noise, that's a lot for comments.

codefriar commented on force-app/main/default/classes/PropertyController.cls:
```diff
@@ -1,107 +1,276 @@
+                AND price__c <= :maxPrice
+                AND beds__c >= :numberBedrooms
+                AND baths__c >= :numberBathrooms
+        ];
```
Not in this case. In this case, it's a COUNT() query.

codefriar commented on force-app/main/default/classes/EinsteinVisionController.cls:
```diff
@@ -1,28 +1,42 @@
+                SELECT Title, VersionData
+                FROM ContentVersion
+                WHERE Title = 'einstein_platform'
+                LIMIT 1
```
This was not one of the queries the CLI scanner identified. Investigating.

codefriar replied:
@pozil - While i'd initially gone through this based on the report the cli scanner team sent - and thus overlooked this one - it turns out querying contentVersion with 'WITH SECURITY_ENFORCED' enabled, throws an error - it's not allowed on contentVersion.

codefriar commented on force-app/main/default/classes/HandlerDashboard.cls:
```diff
@@ -1,30 +1,51 @@
+            FROM Dashboard
+            WHERE Title LIKE :key
+            ORDER BY Title
+            LIMIT 5
```
With Security enforced is not allowed on Dashboard.

codefriar commented on force-app/main/default/classes/HandlerFileUpload.cls:
```diff
@@ -1,22 +1,49 @@
+            ContentDocument doc = [
+                SELECT Id
+                FROM ContentDocument
+                WHERE LatestPublishedVersionId = :v.Id
```
Doesn't apply because the lastPublishedVersionId is a contentVersion id and contentVersion can't use WITH SECURITY_ENFORCED

codefriar commented on force-app/main/default/classes/HandlerFindAccount.cls:
```diff
@@ -1,23 +1,44 @@
+            FROM Account
+            WHERE Name LIKE :key
+            ORDER BY Name
+            LIMIT 5
```
Good catch. I don't know why the cli scanner team didn't pick this up.

Commit:
adjusting a few more queries.

codefriar commented:
@pozil - I did a scan of other queries and pulled my own PMD report locally. I've addressed other queries where WITH SECURITY_ENFORCED makes sense, but there is an entire rabbit hole of other issues here to address. For now, unless this project is re-activated, I'd like to call it done having addressed the issues the CLI scanner team identified.

pozil commented:
Got it, good to merge then.
